core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -240,6 +294,46 @@ public float matchCost() {
             return 42; // that's obvious, right?
         }
     }
+
+    /**
+     * This TwoPhaseIterator marks all nested docs of matching parents as matches as well.
+     */
+    private final class NestedRoutingPartitionedDocIdSetIterator extends RoutingPartitionedDocIdSetIterator {
+        private final BitSet parentDocs;
+        private int nextParent = -1;
+        private boolean nextParentMatches;
+
+        NestedRoutingPartitionedDocIdSetIterator(LeafReader leafReader, Visitor visitor, BitSet parentDocs) {
+            super(leafReader, visitor);
+            this.parentDocs = parentDocs;
+        }
+
+        @Override
+        public boolean matches() throws IOException {
```

Comment:
I was trying to understand why this works, because of the forward iteration here (with nested, we usually seek backwards (`BitSet.prevSetBit(...)`)). So this works because all live doc ids (root docs and nested docs) are evaluated in order.

Reply:
correct, I will leave a comment